File: _1_data_structures/_1_arrays/_40_sum_in_a_matrix.cpp
// https://leetcode.com/problems/sum-in-a-matrix/description/

class Solution {
public:
    int matrixSum(vector<vector<int>>& nums) {
        for(int i=0;i<nums.size();i++) {
            sort(nums[i].begin(), nums[i].end());
        }
        
        long sum = 0;
        
        vector<int> me;
        int s = nums[0].size()-1;
        for(int j=s;j>=0;j--) {
            for(int i=0;i<nums.size();i++){
            me.push_back(nums[i][j]);
        }
            int ele = maxi(me);
            sum+=ele;
            me.clear();
            
        }
        
        return sum;
        
        
        
    }
    
    int maxi(vector<int> &v) {
        int m = INT_MIN;
        for(int i=0;i<v.size();i++) {
            m = max(m, v[i]);
        }
        if(m==INT_MIN) return 0;
        return m;
    }
};

File: _dsa_go_java/_1_data_structures/_1_data_structures_go/_1_arrays/40-sum-in-a-matrix.go
// https://leetcode.com/problems/sum-in-a-matrix/description/

package main

import (
	"fmt"
	"sort"
)

func matrixSum(nums [][]int) int {
	for i := range nums {
		sort.Ints(nums[i])
	}

	sum := 0
	cols := len(nums[0])

	for j := cols - 1; j >= 0; j-- {
		maxVal := 0

		for i := 0; i < len(nums); i++ {
			if nums[i][j] > maxVal {
				maxVal = nums[i][j]
			}
		}

		sum += maxVal
	}

	return sum
}

func main() {
	nums := [][]int{
		{7, 2, 1},
		{6, 4, 2},
		{6, 5, 3},
	}

	fmt.Println(matrixSum(nums))
}

// Overall:

// 👉 O(n * m log m)

// Space:
// 👉 O(1) extra (optimized version)